File: stratoquantum_platform/backend/src/routes/agents.js
// AI Agents routes for Strato Quantum Platform
const express = require('express');
const { body, param, validationResult } = require('express-validator');
const auth = require('../middleware/auth');
const logger = require('../utils/logger');

const router = express.Router();

// Mock agents data
const mockAgents = [
  {
    id: 'financeiro-ai',
    name: 'Agente Financeiro',
    shortName: 'F',
    description: 'Especialista em análise financeira, fluxo de caixa e previsões',
    status: 'online',
    capabilities: ['cash_flow_analysis', 'budget_analysis', 'roi_calculation', 'financial_forecast'],
    workspace: 'financeiro'
  },
  {
    id: 'rh-ai',
    name: 'Agente RH',
    shortName: 'RH',
    description: 'Especialista em gestão de pessoas, recrutamento e desenvolvimento',
    status: 'online',
    capabilities: ['recruitment_analysis', 'performance_analysis', 'culture_analysis', 'training_analysis'],
    workspace: 'rh'
  },
  {
    id: 'tecnologia-ai',
    name: 'Agente Tecnologia',
    shortName: 'T',
    description: 'Especialista em arquitetura, DevOps e segurança',
    status: 'online',
    capabilities: ['architecture_review', 'security_analysis', 'performance_optimization', 'code_review'],
    workspace: 'tecnologia'
  },
  {
    id: 'operacoes-ai',
    name: 'Agente Operações',
    shortName: 'O',
    description: 'Especialista em processos, SLA e gestão de projetos',
    status: 'online',
    capabilities: ['process_optimization', 'sla_monitoring', 'project_management', 'incident_analysis'],
    workspace: 'operacoes'
  },
  {
    id: 'comercial-ai',
    name: 'Agente Comercial',
    shortName: 'C',
    description: 'Especialista em vendas, CRM e pipeline',
    status: 'online',
    capabilities: ['lead_qualification', 'sales_forecast', 'pipeline_analysis', 'conversion_optimization'],
    workspace: 'comercial'
  },
  {
    id: 'produto-ai',
    name: 'Agente Produto',
    shortName: 'P',
    description: 'Especialista em roadmap, features e feedback',
    status: 'online',
    capabilities: ['feature_prioritization', 'user_feedback_analysis', 'roadmap_planning', 'usage_analytics'],
    workspace: 'produto'
  },
  {
    id: 'marketing-ai',
    name: 'Agente Marketing',
    shortName: 'M',
    description: 'Especialista em campanhas, SEO e performance',
    status: 'online',
    capabilities: ['campaign_optimization', 'seo_analysis', 'content_strategy', 'roi_analysis'],
    workspace: 'marketing'
  }
];

// GET /api/agents - List all agents
router.get('/', auth, async (req, res) => {
  try {
    const { workspace, status } = req.query;
    let agents = [...mockAgents];

    // Filter by workspace
    if (workspace) {
      agents = agents.filter(agent => agent.workspace === workspace);
    }

    // Filter by status
    if (status) {
      agents = agents.filter(agent => agent.status === status);
    }

    logger.info(`Retrieved ${agents.length} agents for user ${req.user.id}`);

    res.json({
      success: true,
      data: agents,
      total: agents.length,
      timestamp: new Date().toISOString()
    });
  } catch (error) {
    logger.error('Error retrieving agents:', error);
    res.status(500).json({
      success: false,
      message: 'Erro interno do servidor',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

// GET /api/agents/:agentId - Get specific agent
router.get('/:agentId', auth, [
  param('agentId').notEmpty().withMessage('Agent ID is required')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: 'Dados inválidos',
        errors: errors.array()
      });
    }

    const { agentId } = req.params;
    const agent = mockAgents.find(a => a.id === agentId);

    if (!agent) {
      return res.status(404).json({
        success: false,
        message: 'Agente não encontrado'
      });
    }

    logger.info(`Retrieved agent ${agentId} for user ${req.user.id}`);

    res.json({
      success: true,
      data: agent,
      timestamp: new Date().toISOString()
    });
  } catch (error) {
    logger.error('Error retrieving agent:', error);
    res.status(500).json({
      success: false,
      message: 'Erro interno do servidor',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

// POST /api/agents/:agentId/chat - Send message to agent
router.post('/:agentId/chat', auth, [
  param('agentId').notEmpty().withMessage('Agent ID is required'),
  body('message').notEmpty().withMessage('Message is required'),
  body('conversationId').optional().isString()
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: 'Dados inválidos',
        errors: errors.array()
      });
    }

    const { agentId } = req.params;
    const { message, conversationId } = req.body;

    const agent = mockAgents.find(a => a.id === agentId);
    if (!agent) {
      return res.status(404).json({
        success: false,
        message: 'Agente não encontrado'
      });
    }

    // Simulate agent processing time
    await new Promise(resolve => setTimeout(resolve, 1000 + Math.random() * 2000));

    // Generate mock response based on agent type
    const responses = generateAgentResponse(agent, message);

    const chatResponse = {
      id: `msg_${Date.now()}`,
      agentId,
      conversationId: conversationId || `conv_${Date.now()}`,
      userMessage: message,
      agentResponse: responses.response,
      timestamp: new Date().toISOString(),
      metadata: {
        processingTime: Math.round(1000 + Math.random() * 2000),
        confidence: responses.confidence,
        capabilities_used: responses.capabilities
      }
    };

    logger.info(`Agent ${agentId} processed message for user ${req.user.id}`);

    res.json({
      success: true,
      data: chatResponse,
      timestamp: new Date().toISOString()
    });
  } catch (error) {
    logger.error('Error processing agent chat:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao processar mensagem',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

// GET /api/agents/:agentId/history - Get conversation history
router.get('/:agentId/history', auth, [
  param('agentId').notEmpty().withMessage('Agent ID is required')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: 'Dados inválidos',
        errors: errors.array()
      });
    }

    const { agentId } = req.params;
    const { limit = 50, conversationId } = req.query;

    const agent = mockAgents.find(a => a.id === agentId);
    if (!agent) {
      return res.status(404).json({
        success: false,
        message: 'Agente não encontrado'
      });
    }

    // Mock conversation history
    const history = generateMockHistory(agentId, parseInt(limit));

    res.json({
      success: true,
      data: {
        agentId,
        conversations: history,
        total: history.length
      },
      timestamp: new Date().toISOString()
    });
  } catch (error) {
    logger.error('Error retrieving agent history:', error);
    res.status(500).json({
      success: false,
      message: 'Erro ao recuperar histórico',
      error: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

// Helper functions
function generateAgentResponse(agent, userMessage) {
  const responses = {
    'financeiro-ai': [
      'Analisando seus dados financeiros... Com base no fluxo de caixa atual, recomendo priorizar a cobrança das contas em atraso.',
      'Identifiquei uma oportunidade de otimização nos custos operacionais. Posso detalhar as áreas com maior potencial de economia.',
      'Suas métricas financeiras mostram crescimento saudável. Vou preparar um relatório detalhado com projeções para os próximos trimestres.'
    ],
    'rh-ai': [
      'Analisando o perfil dos candidatos... Encontrei 3 profissionais que se alinham perfeitamente com os requisitos da vaga.',
      'Com base nas métricas de engajamento, sugiro implementar um programa de desenvolvimento para a equipe de tecnologia.',
      'O clima organizacional está positivo. Identifiquei algumas oportunidades para melhorar ainda mais a satisfação dos colaboradores.'
    ],
    'tecnologia-ai': [
      'Revisando a arquitetura atual... Identifiquei alguns pontos de otimização que podem melhorar a performance em 40%.',
      'Analisando os logs de segurança... Tudo está funcionando dentro dos parâmetros normais. Recomendo manter os patches atualizados.',
      'O pipeline de CI/CD está otimizado. Posso sugerir algumas melhorias para reduzir ainda mais o tempo de deploy.'
    ],
    'operacoes-ai': [
      'Analisando os processos operacionais... Identifiquei 2 gargalos que podem ser automatizados para aumentar a eficiência.',
      'Os SLAs estão sendo cumpridos adequadamente. Vou monitorar as métricas e alertar sobre qualquer desvio.',
      'Revisando o status dos projetos... Todos estão no prazo. Posso detalhar o progresso de cada iniciativa.'
    ],
    'comercial-ai': [
      'Analisando o pipeline de vendas... Identifiquei 5 oportunidades com alta probabilidade de fechamento este mês.',
      'Com base no histórico de conversões, recomendo focar nos leads do segmento enterprise para maximizar o ROI.',
      'As métricas comerciais mostram crescimento consistente. Vou preparar uma análise detalhada do funil de vendas.'
    ],
    'produto-ai': [
      'Analisando o feedback dos usuários... A feature mais solicitada é a integração com APIs externas. Posso priorizar no roadmap.',
      'Com base nas métricas de uso, recomendo otimizar a jornada de onboarding para reduzir o churn inicial.',
      'O roadmap está alinhado com as necessidades do mercado. Identifiquei algumas oportunidades de inovação.'
    ],
    'marketing-ai': [
      'Analisando as campanhas ativas... O ROI está 25% acima da meta. Recomendo escalar os canais com melhor performance.',
      'Identifiquei uma oportunidade no segmento B2B que pode aumentar significativamente nossa base de leads qualificados.',
      'As métricas de SEO mostram melhoria consistente. Vou sugerir otimizações para acelerar o crescimento orgânico.'
    ]
  };

  const agentResponses = responses[agent.id] || ['Processando sua solicitação... Retornarei com insights relevantes em breve.'];
  const response = agentResponses[Math.floor(Math.random() * agentResponses.length)];

  return {
    response,
    confidence: 0.85 + Math.random() * 0.15,
    capabilities: agent.capabilities.slice(0, 2) // Use first 2 capabilities
  };
}

function generateMockHistory(agentId, limit) {
  const history = [];
  const now = new Date();

  for (let i = 0; i < Math.min(limit, 10); i++) {
    const timestamp = new Date(now - i * 60 * 60 * 1000); // Each message 1 hour apart
    
    history.push({
      id: `msg_${Date.now()}_${i}`,
      conversationId: `conv_${agentId}_main`,
      userMessage: `Consulta exemplo ${i + 1}`,
      agentResponse: `Resposta do agente para a consulta ${i + 1}`,
      timestamp: timestamp.toISOString(),
      read: i < 5 // Mark first 5 as read
    });
  }

  return history.reverse(); // Most recent first
}

module.exports = router;

File: stratoquantum_agents/api/routes/agents.py
"""
Strato Quantum AI Agents - Agents API Routes
RESTful endpoints for agent management and communication
"""

from typing import Dict, Any, List, Optional
from fastapi import APIRouter, HTTPException, Depends, BackgroundTasks
from pydantic import BaseModel, Field

from ..services.agent_manager import agent_manager
from ..config import settings
from ..utils.logger import setup_logger

logger = setup_logger(__name__)

router = APIRouter()

# Pydantic models
class MessageRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=5000, description="User message to the agent")
    conversation_id: Optional[str] = Field(None, description="Conversation ID for context")
    user_id: Optional[str] = Field(None, description="User ID for tracking")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata")

class MessageResponse(BaseModel):
    success: bool
    agent_id: str
    conversation_id: Optional[str]
    user_id: Optional[str]
    message: str
    response: str
    response_time: float
    timestamp: str
    metadata: Dict[str, Any]

class AgentStatus(BaseModel):
    status: str
    total_conversations: int
    total_messages: int
    average_response_time: float
    last_activity: str
    error_count: int
    uptime: str

class AgentInfo(BaseModel):
    id: str
    name: str
    description: str
    capabilities: List[str]
    status: str
    config: Dict[str, Any]

# Agent information
AGENT_INFO = {
    "financeiro": {
        "id": "financeiro",
        "name": "Agente Financeiro",
        "description": "Especialista em análise financeira, fluxo de caixa e previsões",
        "capabilities": ["cash_flow_analysis", "budget_analysis", "roi_calculation", "financial_forecast"],
        "workspace": "financeiro"
    },
    "rh": {
        "id": "rh",
        "name": "Agente RH",
        "description": "Especialista em gestão de pessoas, recrutamento e desenvolvimento",
        "capabilities": ["recruitment_analysis", "performance_analysis", "culture_analysis", "training_analysis"],
        "workspace": "rh"
    },
    "tecnologia": {
        "id": "tecnologia",
        "name": "Agente Tecnologia",
        "description": "Especialista em arquitetura, DevOps e segurança",
        "capabilities": ["architecture_review", "security_analysis", "performance_optimization", "code_review"],
        "workspace": "tecnologia"
    },
    "operacoes": {
        "id": "operacoes",
        "name": "Agente Operações",
        "description": "Especialista em processos, SLA e gestão de projetos",
        "capabilities": ["process_optimization", "sla_monitoring", "project_management", "incident_analysis"],
        "workspace": "operacoes"
    },
    "comercial": {
        "id": "comercial",
        "name": "Agente Comercial",
        "description": "Especialista em vendas, CRM e pipeline",
        "capabilities": ["lead_qualification", "sales_forecast", "pipeline_analysis", "conversion_optimization"],
        "workspace": "comercial"
    },
    "produto": {
        "id": "produto",
        "name": "Agente Produto",
        "description": "Especialista em roadmap, features e feedback",
        "capabilities": ["feature_prioritization", "user_feedback_analysis", "roadmap_planning", "usage_analytics"],
        "workspace": "produto"
    },
    "marketing": {
        "id": "marketing",
        "name": "Agente Marketing",
        "description": "Especialista em campanhas, SEO e performance",
        "capabilities": ["campaign_optimization", "seo_analysis", "content_strategy", "roi_analysis"],
        "workspace": "marketing"
    }
}

@router.get("/", response_model=List[AgentInfo])
async def list_agents():
    """List all available agents"""
    try:
        agents_status = await agent_manager.get_all_agents_status()
        
        agents_list = []
        for agent_id, info in AGENT_INFO.items():
            agent_status = agents_status.get("agents", {}).get(agent_id, {})
            
            agent_info = AgentInfo(
                id=info["id"],
                name=info["name"],
                description=info["description"],
                capabilities=info["capabilities"],
                status=agent_status.get("status", "unknown"),
                config=settings.get_agent_config(agent_id)
            )
            agents_list.append(agent_info)
        
        return agents_list
        
    except Exception as e:
        logger.error(f"Error listing agents: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve agents list")

@router.get("/{agent_id}", response_model=AgentInfo)
async def get_agent(agent_id: str):
    """Get information about a specific agent"""
    if agent_id not in AGENT_INFO:
        raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")
    
    try:
        agent_status = await agent_manager.get_agent_status(agent_id)
        info = AGENT_INFO[agent_id]
        
        return AgentInfo(
            id=info["id"],
            name=info["name"],
            description=info["description"],
            capabilities=info["capabilities"],
            status=agent_status.get("status", "unknown"),
            config=settings.get_agent_config(agent_id)
        )
        
    except Exception as e:
        logger.error(f"Error getting agent {agent_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to retrieve agent {agent_id}")

@router.get("/{agent_id}/status", response_model=AgentStatus)
async def get_agent_status(agent_id: str):
    """Get detailed status of a specific agent"""
    if agent_id not in AGENT_INFO:
        raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")
    
    try:
        status = await agent_manager.get_agent_status(agent_id)
        
        if "error" in status:
            raise HTTPException(status_code=404, detail=status["error"])
        
        return AgentStatus(**status)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting agent status {agent_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get status for agent {agent_id}")

@router.post("/{agent_id}/chat", response_model=MessageResponse)
async def chat_with_agent(agent_id: str, request: MessageRequest, background_tasks: BackgroundTasks):
    """Send a message to a specific agent"""
    if agent_id not in AGENT_INFO:
        raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")
    
    try:
        # Prepare message data
        message_data = {
            "message": request.message,
            "conversation_id": request.conversation_id,
            "user_id": request.user_id,
            "metadata": request.metadata
        }
        
        # Process message through agent
        response = await agent_manager.process_message(agent_id, message_data)
        
        if not response.get("success", False):
            raise HTTPException(
                status_code=500, 
                detail=f"Agent processing failed: {response.get('error', 'Unknown error')}"
            )
        
        # Log interaction in background
        background_tasks.add_task(log_agent_interaction, agent_id, request.message, response)
        
        return MessageResponse(**response)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error chatting with agent {agent_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to process message for agent {agent_id}")

@router.post("/{agent_id}/restart")
async def restart_agent(agent_id: str):
    """Restart a specific agent"""
    if agent_id not in AGENT_INFO:
        raise HTTPException(status_code=404, detail=f"Agent {agent_id} not found")
    
    try:
        success = await agent_manager.restart_agent(agent_id)
        
        if not success:
            raise HTTPException(status_code=500, detail=f"Failed to restart agent {agent_id}")
        
        return {
            "success": True,
            "message": f"Agent {agent_id} restarted successfully",
            "timestamp": settings.get_timestamp()
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error restarting agent {agent_id}: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to restart agent {agent_id}")

@router.get("/status/all")
async def get_all_agents_status():
    """Get status of all agents"""
    try:
        status = await agent_manager.get_all_agents_status()
        return {
            "success": True,
            "data": status,
            "timestamp": settings.get_timestamp()
        }
        
    except Exception as e:
        logger.error(f"Error getting all agents status: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve agents status")

# Background task functions
async def log_agent_interaction(agent_id: str, message: str, response: Dict[str, Any]):
    """Log agent interaction for analytics"""
    try:
        # This would typically save to database
        logger.info(f"Agent interaction logged: {agent_id} - {len(message)} chars - {response.get('response_time', 0):.2f}s")
    except Exception as e:
        logger.error(f"Failed to log agent interaction: {e}")